core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,332 @@
+          case XMLStreamConstants.START_ELEMENT:
+            String nodeNameStart = reader.getLocalName();
+            boolean validHierarchy = true;
+            currentNodeLevelMap.put(nodeLevel, nodeNameStart);
+
+            //check if node hierarchy matches with expected one.
+            for (int j = nodeLevel; j >= 0; j--) {
+              if (!currentNodeLevelMap.get(j).equals(actualNodeLevelMap.get(j))) {
+                validHierarchy = false;
+              }
+            }
+            //check if node hierarchy is valid and it matches with last node in node path
+            //then start appending tag information to create valid XML Record.
+            if (validHierarchy && nodeNameStart.equals(lastNode)) {
+              appendStartTagInformation(nodeNameStart, xmlRecord);
+              //Set flag for valid node path found
+              nodeFound = true;
+              //set file offset
+              currentKey.set(reader.getLocation().getLineNumber());
+            } else if (nodeFound) {
+              //append all child nodes inside valid node path
+              appendStartTagInformation(nodeNameStart, xmlRecord);
+            }
+            nodeLevel++;
+            break;
+          case XMLStreamConstants.CHARACTERS:
+            if (nodeFound) {
+              xmlRecord.append(reader.getText());
+            }
+            break;
+          case XMLStreamConstants.END_ELEMENT:
+            String nodeNameEnd = reader.getLocalName();
+            if (nodeFound) {
+              //add closing tag
+              xmlRecord.append(OPENING_END_TAG_DELIMITER).append(nodeNameEnd).append(CLOSING_END_TAG_DELIMITER);
+              if (nodeNameEnd.equals(lastNode)) {
+                nodeFound = false;
+                //set flag for XML record is ready to emit.
+                xmlRecordReady = true;
+              }
+            }
+            nodeLevel--;
+            break;
+          case XMLStreamConstants.START_DOCUMENT:
+            break;
+        }
+        if (xmlRecordReady) {
+          currentValue.put(fileName, xmlRecord.toString());
+          return true;
+        }
+      }
+    } catch (XMLStreamException exception) {
+      throw new IllegalArgumentException(exception.getMessage());
+    }
+    actionsAfterEOF();
+    return false;
+  }
+
+  /**
+   * Method to append start tag information along with attributes if any
+   * @param nodeNameStart
+   * @param xmlRecord
+   */
+  private void appendStartTagInformation(String nodeNameStart, StringBuilder xmlRecord) {
+    xmlRecord.append(OPENING_START_TAG_DELIMITER).append(nodeNameStart);
+    int count = reader.getAttributeCount();
+    for (int i = 0; i < count; i++) {
+      xmlRecord.append(" ").append(reader.getAttributeLocalName(i)).append("=\"").append(reader.getAttributeValue(i))
+        .append("\"");
+    }
+    xmlRecord.append(CLOSING_START_TAG_DELIMITER);
+  }
+
+  /**
+   * Method to take actions after EOF reached.
+   * @throws IOException
```

Comment:
if you're not explaining anything about the ``IOException``, no need to mention it in the javadocs